src/ILLink.RoslynAnalyzer/TrimAnalysis/TrimAnalysisVisitor.cs

```diff
@@ -0,0 +1,141 @@
+
+		// Override visitor methods to create tracked values when visiting operations
+		// which reference possibly annotated source locations:
+		// - invocations (for annotated method returns)
+		// - parameters
+		// - 'this' parameter (for annotated methods)
+		// - field reference
+
+		public override MultiValue VisitInvocation (IInvocationOperation operation, StateValue state)
+		{
+			// Base logic takes care of visiting arguments, etc.
+			base.VisitInvocation (operation, state);
+
+			return new MultiValue (new SymbolValue (operation.TargetMethod, isMethodReturn: true));
+		}
+
+		public override MultiValue VisitConversion (IConversionOperation operation, StateValue state)
+		{
+			var value = base.VisitConversion (operation, state);
+
+			if (operation.OperatorMethod != null)
+				return new MultiValue (new SymbolValue (operation.OperatorMethod, isMethodReturn: true));
+
+			return value;
+		}
+
+		public override MultiValue VisitParameterReference (IParameterReferenceOperation paramRef, StateValue state)
+		{
+			return new MultiValue (new SymbolValue (paramRef.Parameter));
+		}
+
+		public override MultiValue VisitInstanceReference (IInstanceReferenceOperation instanceRef, StateValue state)
+		{
+			if (instanceRef.ReferenceKind != InstanceReferenceKind.ContainingTypeInstance)
+				return TopValue;
+
+			// The instance reference operation represents a 'this' or 'base' reference to the containing type,
+			// so we get the annotation from the containing method.
+			// TODO: Check whether the Context.OwningSymbol is the containing type in case we are in a lambda.
+			var value = new MultiValue (new SymbolValue ((IMethodSymbol) Context.OwningSymbol, isMethodReturn: false));
+			return value;
+		}
+
+		public override MultiValue VisitFieldReference (IFieldReferenceOperation fieldRef, StateValue state)
+		{
+			return new MultiValue (new SymbolValue (fieldRef.Field));
+		}
+
+		public override MultiValue VisitTypeOf (ITypeOfOperation typeOfOperation, StateValue state)
+		{
+			// TODO: track known types too!
+
+			if (typeOfOperation.TypeOperand is ITypeParameterSymbol typeParameter)
+				return new MultiValue (new SymbolValue (typeParameter));
+
+			return TopValue;
+		}
+
+		// Override handlers for situations where annotated locations may be involved in reflection access:
+		// - assignments
+		// - arguments passed to method parameters
+		//   this also needs to create the annotated value for parmeters, because they are not represented
+		//   as 'IParameterReferenceOperation' when passing arguments
+		// - instance passed as explicit or implicit receiver to a method invocation
+		//   this also needs to create the annotation for the implicit receiver parameter.
+		// - value returned from a method
+
+		public override void HandleAssignment (MultiValue source, MultiValue target, IOperation operation)
+		{
+			if (target.Equals (TopValue))
+				return;
+
+			TrimAnalysisPatterns.Add (new TrimAnalysisPattern (source, target, operation));
```

Comment:
Perf again here: This will track every assignment, whether it's interesting or not. Ideally we would filter this down to only the cases where there's something interesting happening.

Reply:
Added a comment, like above